torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -410,6 +410,15 @@ void XLAGraphExecutor::SyncTensorsGraph(std::vector<XLATensorPtr>* tensors,
       SyncTensorsGraphInternal(tensors, devices, config, warm_up_cache_only);
   if (wait && async != nullptr && !warm_up_cache_only) {
     async->mwait.Wait();
+    // async->mwait.Wait() will block until the async computation thread to
+    // return but the real device execution might not finish.
+    if (UseVirtualDevice()) {
```

Comment:
Can you rebase on this small refactor? #7796

I need to do something similar, but we should avoid copying `if (UseVirtualDevice())` in multiple places.

Reply:
yea let mer rebase.